### PR TITLE
New ExecutionConfig API

### PR DESCRIPTION
**Pull request by fyrestone**
## What do these changes do?

This PR introduces the `ExecutionConfig` API to execution. With the `ExecutionConfig`, we can pass the cluster params to the execution backend.

- [x] New `ExecutionConfig`, the executor accepts an `ExecutionConfig` instance instead of a bare dict.
- [x] Rename the `task_executor_config` to `execution_config`.
- [x] Remove the `ClusterAPI` usage from Ray executor.
- [x] Move the logic of building band resources to `mars/services/task/execution/utils.py`.
- [x] When using Ray backend, the Mars cluster does not start the workers.

### How to add an API for the task processor (the parts other than execution) ?

The API can be put into the `ExecutionConfig`, some backends can have a different implementation.

### How to add an API for Ray executor (only used in the execution) ?

First, we can put the default value to the `config.yml`:

```yml
task:
  execution_config:
    backend: mars
    ray:
        ray_executor_specific_option: 5
```

Then, we can add an API `get_specific_opiton()` to the `RayExecutionConfig` to read the value of `ray_executor_specific_option` from the `self._execution_config`:

```json
{
  "backend": "ray",
  "ray": {
    "ray_executor_specific_option": 5
  }
}
```

## Related issue number

https://github.com/mars-project/mars/issues/2893

## Check code requirements

- [ ] tests added / passed (if needed)
- [ ] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/services/task/supervisor/manager.py [hunk 1]
```diff
@@ -40,20 +40,20 @@ class TaskConfigurationActor(mo.Actor):
     def __init__(
         self,
         task_conf: Dict[str, Any],
-        task_executor_config: Dict[str, Any],
+        execution_config: Dict[str, Any],
```
Comment (qinxuye) [inline mars/services/task/supervisor/manager.py:43]: Is the type right?

Comment (fyrestone) [reply mars/services/task/supervisor/manager.py:43]: The type here is right. The `TaskExecutor` accepts the execution config in dict type and converts it to the correct `ExecutionConfig` object according to the backend. e.g. The `RayTaskExecutor` will be constructed with a `RayExecutionConfig`.

**Review by qinxuye — DISMISSED** [review]
LGTM

**Review by hekaisheng — APPROVED** [review]
LGTM

**Comment by chaokunyang** [comment]
How do I set default config which only ray backend need but mars doesn't with this API in rayconfig.yml

**Comment by fyrestone** [comment]
> How do I set default config which only ray backend need but mars doesn't with this API in rayconfig.yml

The default config can be put in the `rayconfig.yml` like this,

```yml
task:
  execution_config:
    backend: mars
    ray:
        subtask_max_retries: 5
```
Then we can add some APIs in the `RayExecutionConfig` to read the `subtask_max_retries` value from `self._execution_config`:
```python
{
  "backend": "ray",
  "ray": {
    "subtask_max_retries": 5
  }
}
```

File: mars/deploy/oscar/tests/test_local.py [hunk 2]
```diff
@@ -135,6 +138,48 @@ async def _assert(session_id: str, addr: str, level: StorageLevel):
...
+        return_deploy_band_resources = []
+
+        def _wrap_original_deploy_band_resources(*args, **kwargs):
+            nonlocal return_deploy_band_resources
+            return_deploy_band_resources = original_deploy_band_resources(
+                *args, **kwargs
+            )
+            return return_deploy_band_resources
+
+        config_init.side_effect = original_config_init
```
Comment (chaokunyang) [inline mars/deploy/oscar/tests/test_local.py:163]: Seems `config_init` is not mocked?

File: mars/services/task/execution/api.py [hunk 3]
```diff
@@ -14,12 +14,124 @@
...
+            execution_config.get_execution_config(),
+        )
+        return self
+
+    @property
+    def backend(self) -> str:
+        """The backend from config."""
+        return self._execution_config["backend"]
+
+    def get_execution_config(self) -> Dict:
```
Comment (chaokunyang) [inline mars/services/task/execution/api.py:74]: Using a `property`?

File: mars/services/task/execution/api.py [hunk 4]
```diff
@@ -14,12 +14,124 @@
...
+        """Get the execution config dict."""
+        return self._execution_config
+
+    @abstractmethod
+    def get_deploy_band_resources(self) -> List[Dict[str, Resource]]:
+        """Get the band resources for deployment."""
+
+    @classmethod
+    def from_config(cls, config: Dict, backend: str = None) -> "ExecutionConfig":
```
Comment (chaokunyang) [inline mars/services/task/execution/api.py:83]: This passes the whole mars config, do we really need this? Seems `from_execution_config` is enough.

File: mars/resource.py [hunk 5]
```diff
@@ -391,3 +391,25 @@ def cuda_card_stats() -> List[_cuda_card_stat]:  # pragma: no cover
             )
         )
     return infos
+
+
+def build_band_resources(
```
Comment (chaokunyang) [inline mars/resource.py:396]:
```suggestion
def build_bands_resources(
```

File: mars/services/task/execution/core.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+
+from typing import List, Dict
+from ....resource import Resource, build_band_resources
+
+
+class _CommonPrivateConfigMixin:
+    """This class should ONLY provide the common private APIs for all backend."""
+
+    def _get_band_resources(self) -> List[Dict[str, Resource]]:
```
Comment (chaokunyang) [inline mars/services/task/execution/core.py:23]:
```suggestion
    def _get_bands_resources(self) -> List[Dict[str, Resource]]:
```

File: mars/services/task/execution/api.py [hunk 7]
```diff
@@ -14,12 +14,124 @@
...
+    def backend(self) -> str:
+        """The backend from config."""
+        return self._execution_config["backend"]
+
+    def get_execution_config(self) -> Dict:
+        """Get the execution config dict."""
+        return self._execution_config
+
+    @abstractmethod
+    def get_deploy_band_resources(self) -> List[Dict[str, Resource]]:
```
Comment (chaokunyang) [inline mars/services/task/execution/api.py:79]:
```suggestion
    def get_deploy_bands_resources(self) -> List[Dict[str, Resource]]:
```

Comment (chaokunyang) [inline mars/services/task/execution/api.py:79]: What's the difference between `get_deploy_band_resources` and `get_band_resources`

File: mars/services/task/execution/ray/executor.py [hunk 8]
```diff
@@ -236,9 +236,17 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
             await self._lifecycle_api.incref_tileables(tileable_keys)
 
     async def get_available_band_resources(self) -> Dict[BandType, Resource]:
-        async for bands in self._cluster_api.watch_all_bands():
-            if bands:
-                return bands
+        if self._available_band_resources is None:
```
Comment (chaokunyang) [inline mars/services/task/execution/ray/executor.py:239]:
```suggestion
        if self._available_bands_resources is None:
```

File: mars/services/task/execution/ray/executor.py [hunk 9]
```diff
@@ -236,9 +236,17 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
...
+            band_resources = self._config.get_band_resources()
+            virtual_band_resources = {}
+            idx = 0
+            for band_resource in band_resources:
```
Comment (chaokunyang) [inline mars/services/task/execution/ray/executor.py:243]:
```suggestion
            for band_resource in bands_resources:
```

File: mars/services/task/execution/api.py (hunk 4 above)
Comment (fyrestone) [reply mars/services/task/execution/api.py:83]: This is for merging the execution config part of the mars config: https://github.com/mars-project/mars/pull/2968/files#diff-1b98ea665f30f15485deb7982cdb0fc5e9127cb2225b9197811135ecbaefee73R156

With out this API, we have to access the execution config from dict by some keys, it is not maintainable.

File: mars/resource.py (hunk 5 above)
Comment (fyrestone) [reply mars/resource.py:396]: It's a collection of (band, resource) pair, I think the `build_band_resources` is good enough. Currently, all the names are `band_resources` not `bands_resources`. @qinxuye @hekaisheng @wjsi What do you think?

File: mars/services/task/execution/api.py (hunk 7 above)
Comment (fyrestone) [reply mars/services/task/execution/api.py:79]: The `get_deploy_band_resources` is for the deployment, the comments is here: https://github.com/mars-project/mars/pull/2968/files#diff-28df8dd40f8370801f54fe9d00b8f745446601d87f294d1be220d2afdc3c3acaR80

While, the `get_band_resources` just get the band resources from the config.

File: mars/resource.py (hunk 5 above)
Comment (hekaisheng) [reply mars/resource.py:396]: I think `band_resources` is better for me.

**Review by chaokunyang — DISMISSED** [review]
LGTM

File: mars/services/task/execution/api.py (hunk 3 above)
Comment (fyrestone) [reply mars/services/task/execution/api.py:74]: I think that this API should not be used very often. So, it is a method not a property.

File: mars/resource.py (hunk 5 above)
Comment (wjsi) [inline mars/resource.py:396]: `band` is a concept introduced in Cluster Service. Thus not appropriate to introduce it in `mars.resource`.

File: mars/services/task/execution/core.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from typing import List, Dict
+from ....resource import Resource, build_band_resources
+
+
+class _CommonPrivateConfigMixin:
```
Comment (wjsi) [inline mars/services/task/execution/core.py:20]: A little strange for me as there is no reason to place a separate Mixin for all backends.

Comment (fyrestone) [reply mars/services/task/execution/core.py]: Fixed.

File: mars/resource.py (hunk 5 above)
Comment (fyrestone) [reply mars/resource.py:396]: So, where is the best place for the `build_band_resources`?

Comment (fyrestone) [reply mars/resource.py]: Fixed. Thanks.

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**